staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -396,6 +396,13 @@ func (r *crdHandler) serveResource(w http.ResponseWriter, req *http.Request, req
 		forceWatch := true
 		return handlers.ListResource(storage, storage, requestScope, forceWatch, r.minRequestTimeout)
 	case "create":
+		// we want to track recently created CRDs so that in HA environments we don't have server A allow a create and server B
+		// not have observed the established, so a followup get,update,delete results in a 404. We've observed about 800ms
+		// delay in some CI environments.  Two seconds looks long enough and reasonably short for hot retriers.
+		justCreated := time.Now().Sub(apiextensionshelpers.FindCRDCondition(crd, apiextensionsv1.Established).LastTransitionTime.Time) < 2*time.Second
+		if justCreated {
+			time.Sleep(2 * time.Second)
```

Comment:
> 5. server1 has not observed CRD established for whatever reason and returns a 404

@deads2k could we read a CRD directly from the etcd (not from the cache) on a 404 instead of sleeping?

reading from highly consistent storage is guaranteed to return the same result for all APIs whereas sleeping is not.

Reply:
IMO, if not reading it from etcd here, there may be a performance issue with a large number of CRD creations.

Reply:
> IMO, if not reading it from etcd here, there may be a performance issue with a large number of CRD creations.

I don't see a perf issue.  For two seconds after CRD creation, CRs cannot be created.  Since the process is already asynchronous, clients have to deal failures.  This doesn't even fail, it simply holds the first few.  Over the lifespan of a CRD it doesn't matter.

Reply:
> @deads2k could we read a CRD directly from the etcd (not from the cache) on a 404 instead of sleeping?

Assuming you are talking about server1's delete call, that means that for every call while not established or accepted would result in an additional API call.  That tradeoff seems excessively costly for the problem.

Reply:
I was referring to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L277 unless it can fail in a different place.

Reply:
> I was referring to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L277 unless it can fail in a different place.

A live get for resources that don't exist isn't worth the tradeoff here.

Reply:
I have just realized that it can also fail on

1. CRD is created
2. server0 observes established
3. client creates a CR instance on server1
4. server1 has not observed CRD established for whatever reason and returns a 404
5. client is confused

Reply:
right, CRD creation and CR success is still not race-free. But this is a little better than "CR create success followed by CR 404".